pkg/virt-dhcp/dnsmasq.go

```diff
@@ -0,0 +1,302 @@
+package virtdhcp
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"syscall"
+
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+const (
+	DNSmasqExec = "/usr/sbin/dnsmasq"
+	DHCPPidFile = "/tmp/dhcp.pid"
+)
+
+var DHCPLeaseFile = "/var/run/kubevirt/dhcp.leases"
+var DHCPHostsFile = "/var/run/kubevirt/dhcp.hosts"
+
+type hostsDetails struct {
+	Mac   string
+	Lease int
+}
+
+type DNSmasqInstance struct {
+	hosts    map[string]hostsDetails
+	ipRange  [][]byte
+	confpath string
+	args     []string
+	cmd      *exec.Cmd
+}
+
+func NewDNSmasq() *DNSmasqInstance {
+	dnsMasq := new(DNSmasqInstance)
+	dnsMasq.hosts = make(map[string]hostsDetails)
+	dnsMasq.ipRange = make([][]byte, 2)
+	dnsMasq.args = []string{"-k", "-d", "--strict-order", "--bind-dynamic", "--no-resolv", "--dhcp-no-override", "--dhcp-authoritative",
+		"--except-interface=lo", "--dhcp-hostsfile=" + DHCPHostsFile, "--dhcp-leasefile=" + DHCPLeaseFile,
+		"--pid-file=" + DHCPPidFile}
+	OS = &OSHandler{}
+	return dnsMasq
+}
+
+// implement `Interface` in sort package.
+type sortByteArrays [][]byte
+
+func (b sortByteArrays) Len() int {
+	return len(b)
+}
+
+func (b sortByteArrays) Less(i, j int) bool {
+	// bytes package already implements Comparable for []byte.
+	switch bytes.Compare(b[i], b[j]) {
+	case -1:
+		return true
+	case 0, 1:
+		return false
+	default:
+		return false
+	}
+}
+
+func (b sortByteArrays) Swap(i, j int) {
+	b[j], b[i] = b[i], b[j]
+}
+
+type OSOps interface {
+	IsFileExist(path string) bool
+	getPidFromFile(file string) (int, error)
+	checkProcessExist(pid int) (bool, *os.Process)
+	killProcessIfExist(pid int) error
+	RecreateHostsFile() (*os.File, error)
+	executeCommand(string, []string) (*exec.Cmd, error)
+	readFromFile(file string) ([]byte, error)
+}
+
+type OSHandler struct{}
+
+var OS OSOps
+
+func (o *OSHandler) IsFileExist(path string) bool {
+	_, err := os.Stat(DHCPPidFile)
+	return !os.IsNotExist(err)
+}
+
+func (o *OSHandler) readFromFile(file string) ([]byte, error) {
+	content, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to open file: %s", file)
+		return nil, err
+	}
+	return content, nil
+}
+
+func (o *OSHandler) getPidFromFile(file string) (int, error) {
+	content, err := OS.readFromFile(file)
+	if err != nil {
+		return -1, err
+	}
+	lines := strings.Split(string(content), "\n")
+	pid, _ := strconv.Atoi(lines[0])
+	return pid, nil
+}
+
+func (o *OSHandler) killProcessIfExist(pid int) error {
+	procExist, proc := OS.checkProcessExist(pid)
+	if procExist {
+		killErr := proc.Kill()
+		if killErr != nil {
+			log.Log.Warningf("failed to kill process %d", proc.Pid)
+		}
+	}
+	return nil
+}
+
+func (o *OSHandler) checkProcessExist(pid int) (bool, *os.Process) {
```

Comment:
I know what I'm about to say is going to sound like overkill, but it happened to me once in production and I'll never forget it. It was a nightmare to debug. 

An exited pid can get reclaimed and assigned to another process by the os. If you're in the host pid namespace and you kill a pid in a pid file, you aren't guaranteed the process you killed is the one associated with the process you think it is.

Because of this, it would be a really good ideal to check the /proc/$pid/cmdline to ensure the pid is associated with the process you expect it to be.